### PR TITLE
Location permissions2

### PR DESCRIPTION
**Pull request by esoergel**
Restrict web users' access to locations.
- domain admins should have unrestricted access
- field implementers should have write access to their location and any locations beneath them in the tree.
- they should also be able to see (but not edit) their direct ancestors

http://manage.dimagi.com/default.asp?167182#933759
@emord did most of the work on this
@benrudolph @czue review-buddy pings

### CODE REVIEW

File: corehq/apps/locations/permissions.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            not project.location_restriction_for_users):
+        return (SQLLocation.by_domain(project.name)
+                           .values_list('location_id', flat=True))
+
+    user_loc = user.get_location(project.name)
+    if not user_loc:
+        return []
+
+    return list(user_loc.sql_location.get_descendants(include_self=True)
+                                 .values_list('location_id', flat=True))
```
Comment (dimagimon) [inline corehq/apps/locations/permissions.py]: E127 continuation line over-indented for visual indent

File: corehq/apps/locations/resources/v0_1.py [hunk 2]
```diff
@@ -1,5 +1,8 @@
 from tastypie import fields
 from corehq.apps.locations.models import Location, root_locations
+from corehq.apps.locations.permissions import (user_can_edit_location,
```
Comment (dimagimon) [inline corehq/apps/locations/resources/v0_1.py]: F401 'user_can_edit_location' imported but unused

File: corehq/apps/locations/views.py [hunk 3]
```diff
@@ -31,17 +31,26 @@
 from corehq.util import reverse, get_document_or_404
 from custom.openlmis.tasks import bootstrap_domain_task
 
-from .models import Location, LocationType, SQLLocation
+from .permissions import (locations_access_required, is_locations_admin,
+                          can_edit_location, can_edit_location_types)
+from .models import Location, LocationType, SQLLocation, all_locations
```
Comment (dimagimon) [inline corehq/apps/locations/views.py]: F401 'all_locations' imported but unused

File: corehq/apps/locations/views.py [hunk 4]
```diff
@@ -68,13 +77,16 @@ def show_inactive(self):
...
         return {
             'selected_id': selected_id,
             'locations': load_locs_json(
-                self.domain, selected_id, self.show_inactive
+                self.domain, selected_id, self.show_inactive, self.request.couch_user
             ),
             'show_inactive': self.show_inactive,
-            'has_location_types': has_location_types
+            'has_location_types': has_location_types,
+            'can_edit_root': (not loc_restricted or
```
Comment (dimagimon) [inline corehq/apps/locations/views.py]: W291 trailing whitespace

**Comment by emord** [comment]
Looks like this PR is half my commits so cc enchantress: @czue

**Comment by czue** [comment]
is there a spec or fogbugz ticket for what this does? if not maybe you could write something in the PR description?

File: corehq/apps/locations/permissions.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+from .models import SQLLocation
+from corehq.apps.domain.decorators import (login_and_domain_required,
+                                           domain_admin_required)
+from corehq.util.quickcache import quickcache
+
+
+def locations_access_required(view_fn):
+    """
+    Decorator controlling domain-level access to locations.
+    Mostly a placeholder, soon this will also check for standard plan
```
Comment (czue) [inline corehq/apps/locations/permissions.py]: any reason the plan check didn't get into this PR?

Comment (esoergel) [inline corehq/apps/locations/permissions.py]: I wanted to do a bit of QA on that, and didn't want this to be blocked by that.

**Comment by emord** [comment]
@czue ticket is here: http://manage.dimagi.com/default.asp?167182#933759

**Comment by esoergel** [comment]
This is g2m (pending review) as far as I'm concerned.
ping @benrudolph

**Comment by benrudolph** [comment]
looking now, sorry slipped from my radar

File: corehq/apps/locations/resources/v0_1.py [hunk 6]
```diff
@@ -20,13 +24,22 @@ def obj_get(self, bundle, **kwargs):
...
+            locs = root_locations(domain)
+        else:
             parent = get_object_or_not_exist(Location, parent_id, domain)
-            return parent.sql_location.child_locations(include_archive_ancestors=include_inactive)
+            locs = parent.sql_location.child_locations(include_archive_ancestors=include_inactive)
+
+        return [child for child in locs if child.location_id in viewable]
 
-        return root_locations(domain)
+    def dehydrate_can_edit(self, bundle):
```
Comment (benrudolph) [inline corehq/apps/locations/resources/v0_1.py]: where is this used? 
small: also not sure what `dehydrate` means in this context, maybe a comment?

Comment (emord) [inline corehq/apps/locations/resources/v0_1.py]: http://django-tastypie.readthedocs.org/en/latest/resources.html#the-dehydrate-cycle it's a tastypie thing

Comment (esoergel) [inline corehq/apps/locations/resources/v0_1.py]: This is a tastypie (api library) thing.  You know how django forms have magic `clean_<field name>` methods?  Tastypie has something similar for transforming fields.

File: corehq/apps/users/models.py [hunk 7]
```diff
@@ -2256,6 +2256,14 @@ def get_dimagi_emails_by_domain(cls, domain):
             if user_doc['email'].endswith('@dimagi.com'):
                 yield user_doc['email']
 
+    def get_location_id(self, domain):
+        return getattr(self.get_domain_membership(domain), 'location_id', None)
+
+    def get_location(self, domain):
```
Comment (benrudolph) [inline corehq/apps/users/models.py]: minor: this gets called quite a bit, may be a good case for memoize or quickcache

File: corehq/apps/locations/resources/v0_1.py (hunk 6 above)
Comment (benrudolph) [inline corehq/apps/locations/resources/v0_1.py]: gotcha ok

**Comment by benrudolph** [comment]
nothing blocking. going to merge

**Comment by benrudolph** [comment]
also would be nice to have a test. i know the `can_edit` thing was a pretty big nightmare in the app manager (slightly different context though so maybe this'll be easier to maintain)

**Pull request merged**